onboard/catkin_ws/src/controls/src/drc_pid.cpp

```diff
@@ -0,0 +1,139 @@
+    this->derivative_type = derivative_type;
+    this->error_ramp_rate = error_ramp_rate;
+    this->pid_gains = pid_gains;
+}
+
+void PID::set_pid_gain(const PIDGainTypesEnum &pid_gain_type, const double &value) {
+    // Set gains
+    this->pid_gains[pid_gain_type] = value;
+}
+
+const PIDGainsMap &PID::get_pid_gains() const { return pid_gains; }
+
+void PID::reset() {
+    // Reset integral to zero
+    integral = 0.0;
+}
+
+double PID::second_order_butterworth(const std::array<double, 3> &values,
+                                     const std::array<double, 3> &filtered_values) {
+    // Reference: Julius O. Smith III, Intro. to Digital Filters With Audio Applications.
+    // See https://ccrma.stanford.edu/~jos/filters/Example_Second_Order_Butterworth_Lowpass.html for a description of
+    // the math. Implementation adapted from https://bitbucket.org/AndyZe/pid/src/master/src/pid.cpp#lines-271.
+    return (1 / (1 + (c * c) + (1.414 * c))) *
+           (values.at(2) + (2 * values.at(1)) + values.at(0) - ((c * c) - (1.414 * c) + 1) * filtered_values.at(2) -
+            ((-2 * c * c) + 2) * filtered_values.at(1));
+}
+
+double PID::run(double error, double delta_time, PIDInfo &info, double provided_derivative) {
+    // If there are validation errors, an exception is not thrown to maintain continuous operation of controls in the
+    // event of a temporary error. Instead, the function returns 0 and prints an error message.
+
+    // If pid_gains is not valid, return 0 and print error
+    if (!ControlsUtils::pid_gains_map_valid(pid_gains)) {
+        ROS_ERROR("PID run loop error: PID gains map is invalid.");
+        return 0;
+    }
+
+    // If delta time is nonpositive, return 0 and print error
+    if (delta_time <= 0) {
+        ROS_WARN("Delta time must be positive.");
+        return 0;
+    }
+
+    // Update errors
+    // Don't need to deal with angular discontinuities as angular errors will always be in range [-pi, pi]
+    errors.at(2) = errors.at(1);
+    errors.at(1) = errors.at(0);
+
+    // Clip error to be within error ramp rate
+    double max_error_change = error_ramp_rate * delta_time;
+    errors.at(0) = ControlsUtils::clip(error, errors.at(1) - max_error_change, errors.at(1) + max_error_change);
+
+    // Update filtered errors
+    filtered_errors.at(2) = filtered_errors.at(1);
+    filtered_errors.at(1) = filtered_errors.at(0);
+    filtered_errors.at(0) = second_order_butterworth(errors, filtered_errors);
+
+    // Update integral
+    integral += error * delta_time;
+
+    // Integral windup protection against sudden setpoint changes or unactuatable states
+    integral = ControlsUtils::clip(integral, -integral_clamp, integral_clamp);
+
+    // Update derivatives
+    derivs.at(2) = derivs.at(1);
+    derivs.at(1) = derivs.at(0);
+
+    // Calculate derivative
+    double calculated_derivative = (errors.at(0) - errors.at(1)) / delta_time;
+
+    // Update derivative based on derivative type
```

Comment:
Since `derivative_type` is initialized at node creation (when reading the robot config file), and is not changed by any method, we only need to do `derivative_type` validation once at `controls` node creation or `PID` object creation time and only two cases are needed here.

Reply:
This is technically true. However, it's better to be explicit here about how each derivative type affects the behavior of this function. If another derivative type is added in the future and this code is not updated, then it will be easier to debug as the error will be printed.